### PR TITLE
[WIP] Add Python Support to On-Demand Feature Views

### PR DESCRIPTION
**Pull request by Rostifar**
# Add Python Support to On-Demand Feature Views 

## Syntax 
Mirrors what Tecton does, which is a pretty simple approach: 
```
@on_demand_feature_view(
    schema=[
        Field(name="days_since_last_payment_failure", dtype=Float64),
    ],
    sources=[
        input_request_datetime,
        user_last_payment_failure,
    ],
    mode="python"
    tags={},
    owner="ml-data@affirm.com",
)
def time_since_last_payment_failure(inputs: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "days_since_last_payment_failure": (now() - inputs["last_payment_failure"]).days() if inputs["days_since_last_payment_failure"] else None, 
    }
```

## TODO 

- [ ] Some additional changes in `feature_store.py`
- [ ] Determine how to handle duplicate feature names in sources (optional)

### CODE REVIEW

File: sdk/python/feast/feature_store.py [hunk 1]
```diff
@@ -1554,6 +1554,28 @@ def write_to_offline_store(
         provider = self._get_provider()
         provider.ingest_df_to_offline_store(feature_view, table)
 
+    @log_exceptions_and_usage
```
Comment (Rostifar) [inline sdk/python/feast/feature_store.py:1557]: This change is really unrelated to the main point of this PR.

File: protos/feast/core/OnDemandFeatureView.proto [hunk 2]
```diff
@@ -58,6 +58,9 @@ message OnDemandFeatureViewSpec {
 
     // Owner of the on demand feature view.
     string owner = 8;
+
+    // Backend used to handle on-demand transformations. One of python or pandas.
+    string mode = 9;
```
Comment (franciscojavierarceo) [inline protos/feast/core/OnDemandFeatureView.proto:63]: probably should be enum right?

```
enum Mode {
  python = 0;
  pandas = 1;
}

message Msg {
  optional Mode mode = 1;
}
```

The annoying thing we'll have to do is map those from string to integers

File: sdk/python/feast/embedded_go/online_features_service.py [hunk 3]
```diff
@@ -252,9 +252,12 @@ def transformation_callback(
     # the typeguard requirement.
     full_feature_names = bool(full_feature_names)
 
+    if odfv.mode == 'python':
+        raise Exception(f'OnDemandFeatureView mode "python" not supported by EmbeddedOnlineFeatureServer.')
```
Comment (franciscojavierarceo) [inline sdk/python/feast/embedded_go/online_features_service.py:256]: Don't think this needs to be f string

File: sdk/python/feast/on_demand_feature_view.py [hunk 4]
```diff
@@ -428,6 +504,7 @@ def on_demand_feature_view(
             FeatureViewProjection,
         ]
     ],
+    mode: str = "python",
```
Comment (franciscojavierarceo) [inline sdk/python/feast/on_demand_feature_view.py:507]: shouldn't default be pandas to be backwards compatible?

**Review by franciscojavierarceo — CHANGES_REQUESTED** [review]
This is awesome. Thanks for getting it out so quickly too. 

Can you submit this one back to Feast when it's ready?

File: protos/feast/core/OnDemandFeatureView.proto (hunk 2 above)
Comment (Rostifar) [reply protos/feast/core/OnDemandFeatureView.proto:63]: ^ string -> enum
